pandas_gbq/core/sample.py

```diff
@@ -0,0 +1,286 @@
+
+    return _TYPE_SIZES.get(field_type.upper(), _UNKNOWN_TYPE_SIZE_ESTIMATE)
+
+
+def _estimate_row_bytes(fields: Sequence[google.cloud.bigquery.SchemaField]) -> int:
+    """
+    Estimates the logical row size in bytes for a list of BigQuery SchemaField objects,
+    using the provided data type size chart and assuming 1MB for all STRING and BYTES
+    fields.
+
+    Args:
+        schema_fields: A list of google.cloud.bigquery.SchemaField objects
+                       representing the table schema.
+
+    Returns:
+        An integer representing the estimated total row size in logical bytes.
+    """
+    total_size = min(
+        _MAX_ROW_BYTES,
+        sum(_estimate_field_bytes(field) for field in fields),
+    )
+    return total_size
+
+
+def _sample_with_tablesample(
+    table: google.cloud.bigquery.Table,
+    *,
+    bqclient: google.cloud.bigquery.Client,
+    proportion: float,
+    target_row_count: int,
+    progress_bar_type: Optional[str] = None,
+    use_bqstorage_api: bool = True,
+) -> Optional[pandas.DataFrame]:
+    query = f"""
+    SELECT *
+    FROM `{table.project}.{table.dataset_id}.{table.table_id}`
+    TABLESAMPLE SYSTEM ({float(proportion) * 100.0} PERCENT)
+    ORDER BY RAND() DESC
+    LIMIT {int(target_row_count)};
+    """
+    rows = bqclient.query_and_wait(query)
+    return pandas_gbq.core.read.download_results(
+        rows,
+        bqclient=bqclient,
+        progress_bar_type=progress_bar_type,
+        warn_on_large_results=False,
+        max_results=None,
+        user_dtypes=None,
+        use_bqstorage_api=use_bqstorage_api,
+    )
+
+
+def _sample_with_limit(
+    table: google.cloud.bigquery.Table,
+    *,
+    bqclient: google.cloud.bigquery.Client,
+    target_row_count: int,
+    progress_bar_type: Optional[str] = None,
+    use_bqstorage_api: bool = True,
+) -> Optional[pandas.DataFrame]:
+    query = f"""
+    SELECT *
+    FROM `{table.project}.{table.dataset_id}.{table.table_id}`
+    ORDER BY RAND() DESC
+    LIMIT {int(target_row_count)};
+    """
+    rows = bqclient.query_and_wait(query)
+    return pandas_gbq.core.read.download_results(
+        rows,
+        bqclient=bqclient,
+        progress_bar_type=progress_bar_type,
+        warn_on_large_results=False,
+        max_results=None,
+        user_dtypes=None,
+        use_bqstorage_api=use_bqstorage_api,
+    )
+
+
+def sample(
+    table_id: str,
+    *,
+    target_mb: Optional[int] = None,
+    credentials: Optional[google.oauth2.credentials.Credentials] = None,
+    billing_project_id: Optional[str] = None,
+    progress_bar_type: Optional[str] = None,
+    use_bqstorage_api: bool = True,
+) -> Optional[pandas.DataFrame]:
+    """Sample a BigQuery table, attempting to limit the amount of data read.
+
+    This function attempts to sample a BigQuery table to a target size in
+    memory. It prioritizes methods that minimize data scanned and downloaded.
+
+    The sampling strategy is as follows:
```

Comment:
Add two notes: target size is based on estimated row size and may under or overshoot, especially for table types that don't provide estimated size in bytes such as VIEWs.

Also, TABLESAMPLE is a biased sample, based on how blocks are stored internally. Link to https://docs.cloud.google.com/bigquery/docs/table-sampling